app/plugin_api.go

```diff
@@ -655,6 +657,20 @@ func (api *PluginAPI) GetPluginStatus(id string) (*model.PluginStatus, *model.Ap
 	return api.app.GetPluginStatus(id)
 }
 
+func (api *PluginAPI) UploadPlugin(file io.Reader, replace bool) (*model.Manifest, *model.AppError) {
+	if !*api.app.Config().PluginSettings.Enable || !*api.app.Config().PluginSettings.EnableUploads {
+		return nil, model.NewAppError("uploadPlugin", "app.plugin.upload_disabled.app_error", nil, "", http.StatusNotImplemented)
+	}
+
+	fileBuffer, err := ioutil.ReadAll(file)
```

Comment:
What is the purpose of reading the file into the buffer here?

Reply:
As per ticket specifications UploadPlugin method should take an io.Reader. InstallPluin api require a io.ReadSeeker. io.Reader does not implement seek(). To make this happen file needs to be read completely. So reading file in memory and creating a ReadSeeker using bytes.NewReader

Reply:
Ok That make sense.